internal/worker/concurrency_auto_scaler_test.go

```diff
@@ -0,0 +1,323 @@
+			},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			defer goleak.VerifyNone(t)
+			core, obs := observer.New(zap.DebugLevel)
+			logger := zap.New(core)
+			clock := clockwork.NewFakeClockAt(time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC))
+			scaler := createTestConcurrencyAutoScaler(t, logger, clock)
+
+			// mock poller every 1 tick time
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				clock.Sleep(testTickTime / 2) // poll delay by 0.5 unit of time to avoid test flakiness
+				for _, hint := range tt.pollAutoConfigHint {
+					t.Log("hint process time: ", clock.Now().Format(testTimeFormat))
+					scaler.ProcessPollerHint(hint)
+					clock.Sleep(testTickTime)
+				}
+			}()
+
+			scaler.Start()
+			clock.BlockUntil(2)
+
+			// advance clock
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for i := 0; i < len(tt.pollAutoConfigHint)*2+1; i++ {
+					clock.Advance(testTickTime / 2)
+					time.Sleep(10 * time.Millisecond) // process non-time logic
+				}
+			}()
+
+			wg.Wait()
+			scaler.Stop()
+
+			var actualEvents []eventLog
+			for _, event := range obs.FilterMessage(autoScalerEventLogMsg).All() {
+				if event.ContextMap()["event"] != autoScalerEventMetrics {
+					t.Log("event: ", event.ContextMap())
+					actualEvents = append(actualEvents, eventLog{
+						eventType:   autoScalerEvent(event.ContextMap()["event"].(string)),
+						enabled:     event.ContextMap()["enabled"].(bool),
+						pollerQuota: event.ContextMap()["poller_quota"].(int64),
+						time:        event.ContextMap()["time"].(time.Time).Format(testTimeFormat),
+					})
+				}
+			}
+			assert.ElementsMatch(t, tt.expectedEvents, actualEvents)
+		})
+	}
+}
+
+func TestRollingAverage(t *testing.T) {
+	for _, tt := range []struct {
+		name         string
+		cap          int
+		addGoroutine int
+		input        []float64
+		expected     []float64
+	}{
+		{
+			"cap is 0",
+			0,
+			5,
+			[]float64{1, 2, 3, 4, 5, 6, 7},
+			[]float64{0, 0, 0, 0, 0, 0, 0},
+		},
+		{
+			"cap is 1",
+			1,
+			5,
+			[]float64{1, 2, 3, 4, 5, 6, 7},
+			[]float64{1, 2, 3, 4, 5, 6, 7},
+		},
+		{
+			"cap is 2",
+			2,
+			5,
+			[]float64{1, 2, 3, 4, 5, 6, 7},
+			[]float64{1, 1.5, 2.5, 3.5, 4.5, 5.5, 6.5},
+		},
+		{
+			"cap is 3",
+			3,
+			5,
+			[]float64{1, 2, 3, 4, 5, 6, 7},
+			[]float64{1, 1.5, 2, 3, 4, 5, 6},
+		},
+		{
+			"cap is 4",
+			4,
+			5,
+			[]float64{1, 2, 3, 4, 5, 6, 7},
+			[]float64{1, 1.5, 2, 2.5, 3.5, 4.5, 5.5},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			defer goleak.VerifyNone(t)
+			r := newRollingAverage(tt.cap)
+
+			doneC := make(chan struct{})
+
+			inputChan := make(chan float64)
+			var wg sync.WaitGroup
+			for i := 0; i < tt.addGoroutine; i++ {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					for v := range inputChan {
+						r.Add(v)
+						doneC <- struct{}{}
+					}
+				}()
+			}
+
+			for i := range tt.input {
+				inputChan <- tt.input[i]
+				<-doneC
```

Comment:
this is serializing the whole thing, so it doesn't really test anything concurrently - you can even remove the mutexes and it'll still pass the race detector.

what's the goal with the concurrency?  it's how it's used so 👍 concurrency in tests is good, but this has artificially nerfed it.